### PR TITLE
[FBW] Avoid duplicated donwloads, create networks and get hostname

### PR DESCRIPTION
**Pull request by gmarcos87**
## Avoid duplicated downloads
Before it downloaded all the configurations from the associated devices each time it configured the network. Now it does it once per mac address.

## Get remote hostname
Obtains the remote hostname the moment it gets the configurations and saves the data in the file.

## Set username in create and join network functions
The user can input a hostname so that it changes the moment the new network configuration is set.

### CODE REVIEW

File: packages/ubus-lime-fbw/files/www/cgi-bin/hostname [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+echo "Status: 200"
+echo "Content-type: text/plain;"
+echo "Cache-control: max-age=0, no-cache;"
+echo "Access-Control-Allow-Origin: *"
+echo "Access-Control-Allow-Methods: GET;"
+echo "Access-Control-Allow-Headers: X-PINGOTHER, Content-Type;"
+echo
+echo $HOSTNAME
```
Comment (spiccinini) [inline packages/ubus-lime-fbw/files/www/cgi-bin/hostname]: Just a thought, in the future if we start adding more of this we may refactor into something more maintainable than than bash/cgi.

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 2]
```diff
@@ -219,22 +250,31 @@ function read_configs()
...
     utils.execute("/rom/etc/uci-defaults/91_lime-config")
-    utils.execute("rm /etc/first_run")
-    os.execute("(( /usr/bin/lime-config && /usr/bin/lime-apply && reboot 0<&- &>/dev/null &) &)")
-    return { configs = configs }
+    uci_cursor:set("lime", 'system', 'hostname', hostname)
+    uci_cursor:commit('lime')
+    -- Start sharing lime-defaults
+    -- Apply new configuration
+    os.execute("/usr/bin/lime-config")
+    os.execute("/usr/bin/lime-apply")
```
Comment (G10h4ck) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: Why is lime-apply needed if we reboot just after?

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 3]
```diff
@@ -219,22 +250,31 @@ function read_configs()
...
+    -- Start sharing lime-defaults and reboot
+    share_defualts()
+    remove_lock_file()
+    os.execute("reboot 0")
```
Comment (G10h4ck) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: what the 0 after reboot does?

File: packages/ubus-lime-fbw/Makefile [hunk 4]
```diff
@@ -25,6 +25,7 @@ define Package/$(PKG_NAME)/install
 	$(CP) ./files/* $(1)/
 	@chmod a+x $(1)/usr/libexec/daemon/lime-fbw
 	@chmod a+x $(1)/etc/init.d/ubus-lime-fbw
+  @chmod a+x $(1)/www/cgi-bin/hostname
```
Comment (G10h4ck) [inline packages/ubus-lime-fbw/Makefile]: AFAIR  some makefile macro should be used instead con plain chmod to install executables @spiccinini do you remember if it is like that?

File: packages/ubus-lime-fbw/files/www/cgi-bin/hostname (hunk 1 above)
Comment (G10h4ck) [inline packages/ubus-lime-fbw/files/www/cgi-bin/hostname]: missing new line at the end of the file

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua [hunk 5]
```diff
@@ -153,4 +153,8 @@ function utils.append_network(dev)
     end
 end
 
+function utils.filter_alredy_scanned(hosts, results)
+    return ft.filter(function(host) return results[host.host] == nil end, hosts)
+end
+
 return utils
```
Comment (G10h4ck) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: missing new line at the end of file

File: packages/ubus-lime-fbw/Makefile (hunk 4 above)
Comment (spiccinini) [reply packages/ubus-lime-fbw/Makefile]: Yes, `INSTALL_BIN` macro does `chmod +x` (equivalent to `chdmo a+x`)

Comment (spiccinini) [reply packages/ubus-lime-fbw/Makefile]: This package is using "the trick" using `CP ./files/*` so using `INSTALL_* ` macros goes against that practicity. I don't know if there is a preferred way. I see mixed uses in other packages (from openwrt and libremesh).

**Comment by spiccinini** [comment]
This fixes #101  ?

File: packages/ubus-lime-fbw/files/www/cgi-bin/hostname (hunk 1 above)
Comment (gmarcos87) [reply packages/ubus-lime-fbw/files/www/cgi-bin/hostname]: Yeah, we can do a lot more about it. For now I would leave it at that and see if I could find a more general solution.

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 2 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: done in 8dc3f5ee3d782a71d0149ba6cf2095b4d86a6f92

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 3 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: fix typo in 50582c1

File: packages/ubus-lime-fbw/files/www/cgi-bin/hostname (hunk 1 above)
Comment (gmarcos87) [reply packages/ubus-lime-fbw/files/www/cgi-bin/hostname]: fixed in 42b9ba9c7cdc476f4e4c890b606adda409aed1ab

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua (hunk 5 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: fixed in 42b9ba9

**Comment by gmarcos87** [comment]
yes @spiccinini close https://github.com/libremesh/lime-packages/issues/101

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua (hunk 5 above)
Comment (G10h4ck) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: Still I see this missing new line at EOF

Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: ready, Im thinking on adapting this script  https://gist.github.com/johnjohndoe/4024222 :)

**Pull request merged**